File: apps/ios/BrickBack/Features/Catalog/SetPartsScreen.swift
import SwiftUI
import BrickBackKit

/// `.setParts` — the unique (part, colour) lines that make up a set, read from the catalog.
/// Preview only (no counts); counting happens after "Start sorting" snapshots the set locally.
/// Port of `set_parts_screen.dart`. Parts are sorted by colour, then name.
struct SetPartsScreen: View {
    @Environment(AppEnvironment.self) private var env
    @Environment(\.activeRouter) private var activeRouter
    let itemId: Int

    @State private var state: LoadState<[ExpandedPart]> = .loading

    private var router: Router { activeRouter ?? env.homeRouter }

    /// On iOS 18+ this screen rides the native nav bar (back chevron + inline title), matching the
    /// set-detail screen it's pushed from and the rest of the catalog family. The iOS 17 legacy
    /// pushed flow has no reliable native bar, so it keeps its `ScreenHeader`.
    private var systemProvidesBack: Bool {
        if #available(iOS 18.0, *) { return true }
        return activeRouter === env.searchRouter
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            if !systemProvidesBack {
                ScreenHeader(L.uniqueParts, onBack: { router.pop() })
            }
            switch state {
            case .idle, .loading:
                ProgressView().tint(AppColors.primary).frame(maxWidth: .infinity, maxHeight: .infinity)
            case .failed(let message):
                EmptyState(title: L.partsCouldntLoad, message: message, icon: "exclamationmark.triangle")
            case .loaded(let parts):
                if parts.isEmpty {
                    EmptyState(
                        title: L.partsEmptyTitle,
                        message: L.partsEmptyMessage,
                        icon: "square.grid.2x2"
                    )
                } else {
                    list(sorted(parts))
                }
            }
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .topLeading)
        .background(AppColors.canvas)
        // Only shown when the native bar is present (iOS 18+); harmless where it's hidden.
        .navigationTitle(L.uniqueParts)
        .navigationBarTitleDisplayMode(.inline)
        .task(id: itemId) {
            state = .loading
            do {
                let parts = try await env.services.catalog.expandSetParts(itemId)
                if !Task.isCancelled { state = .loaded(parts) }
            } catch {
                if !Task.isCancelled { state = .failed("\(error)") }
            }
        }
    }

    private func sorted(_ parts: [ExpandedPart]) -> [ExpandedPart] {
        parts.sorted { a, b in
            let ca = a.colorName ?? "~", cb = b.colorName ?? "~"
            if ca != cb { return ca < cb }
            return a.partName < b.partName
        }
    }

    private func list(_ parts: [ExpandedPart]) -> some View {
        ScrollView {
            LazyVStack(alignment: .leading, spacing: AppSpacing.s8) {
                Text(uniquePartsCountLabel(parts.count))
                    .font(AppText.caption).foregroundStyle(AppColors.inkSoft)
                    .padding(.bottom, AppSpacing.s4)
                ForEach(parts, id: \.key) { part in PartRow(part: part) }
            }
            .padding(.horizontal, AppSpacing.screen)
            .padding(.top, AppSpacing.s4)
            .padding(.bottom, AppSpacing.s24)
        }
    }
}

/// A catalog part line: thumbnail, name, colour swatch + colour · part-num, ×qty. Shared shape.
struct PartRow: View {
    let part: ExpandedPart

    private var sub: String {
        var parts = [part.colorName ?? L.unknownColor]
        if let num = part.partNum, !num.isEmpty { parts.append(num) }
        return parts.joined(separator: " · ")
    }

    var body: some View {
        AppCard(padding: AppSpacing.s12) {
            HStack(spacing: AppSpacing.s12) {
                SetThumb(imageUrl: part.imageUrl, size: 44)
                VStack(alignment: .leading, spacing: 2) {
                    Text(part.partName).font(AppText.body).foregroundStyle(AppColors.ink).lineLimit(1)
                    HStack(spacing: AppSpacing.s4) {
                        Circle()
                            .fill(swatchColor(part.colorRgb))
                            .frame(width: 10, height: 10)
                            .overlay(Circle().stroke(AppColors.line, lineWidth: 1))
                        Text(sub).font(AppText.caption).foregroundStyle(AppColors.inkSoft).lineLimit(1)
                    }
                }
                Spacer(minLength: AppSpacing.s8)
                Text("×\(part.neededQty)").font(AppText.label).foregroundStyle(AppColors.ink)
            }
        }
    }
}

/// "1 unique part" / "N unique parts" — the Dart `uniquePartsCount` plural.
func uniquePartsCountLabel(_ count: Int) -> String {
    return L.uniquePartsCount(count)
}

File: apps/mobile/lib/features/catalog/set_parts_screen.dart
import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

import '../../l10n/l10n.dart';
import '../../theme/app_theme.dart';
import '../../widgets/primitives.dart';
import '../rebuild/rebuild_models.dart';
import 'catalog_repository.dart';

/// `/set/:id/parts` — the unique (part, colour) lines that make up a set, read
/// from the catalog. Preview only (no counts); counting happens after "Start
/// sorting" snapshots the set locally.
class SetPartsScreen extends ConsumerWidget {
  const SetPartsScreen({super.key, required this.itemId});

  final int itemId;

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    final c = BrickColors.of(context);
    final parts = ref.watch(setPartsProvider(itemId));
    return Scaffold(
      body: SafeArea(
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.stretch,
          children: [
            ScreenHeader(context.l10n.uniqueParts, onBack: () => Navigator.of(context).maybePop()),
            Expanded(
              child: parts.when(
                loading: () =>
                    Center(child: CircularProgressIndicator(color: c.primary)),
                error: (e, _) => EmptyState(
                  icon: Icons.error_outline,
                  title: context.l10n.partsCouldntLoad,
                  message: '$e',
                ),
                data: (list) {
                  if (list.isEmpty) {
                    return EmptyState(
                      icon: Icons.grid_view_outlined,
                      title: context.l10n.partsEmptyTitle,
                      message: context.l10n.partsEmptyMessage,
                    );
                  }
                  final sorted = [...list]..sort((a, b) {
                    final c = (a.colorName ?? '~').compareTo(b.colorName ?? '~');
                    return c != 0 ? c : a.partName.compareTo(b.partName);
                  });
                  return ListView.separated(
                    padding: const EdgeInsets.fromLTRB(
                        AppSpacing.screen, AppSpacing.s4, AppSpacing.screen, AppSpacing.s24),
                    itemCount: sorted.length + 1,
                    separatorBuilder: (_, _) => const SizedBox(height: AppSpacing.s8),
                    itemBuilder: (context, i) {
                      if (i == 0) {
                        return Padding(
                          padding: const EdgeInsets.only(bottom: AppSpacing.s4),
                          child: Text(context.l10n.uniquePartsCount(sorted.length),
                              style: AppText.caption),
                        );
                      }
                      return _PartRow(part: sorted[i - 1]);
                    },
                  );
                },
              ),
            ),
          ],
        ),
      ),
    );
  }
}

class _PartRow extends StatelessWidget {
  const _PartRow({required this.part});
  final ExpandedPart part;

  @override
  Widget build(BuildContext context) {
    final c = BrickColors.of(context);
    final sub = [
      part.colorName ?? context.l10n.colorUnknown,
      if (part.partNum != null) part.partNum!,
    ].join(' · ');
    return AppCard(
      padding: const EdgeInsets.all(AppSpacing.s12),
      child: Row(
        children: [
          SetThumb(imageUrl: part.imageUrl, size: 44),
          const SizedBox(width: AppSpacing.s12),
          Expanded(
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: [
                Text(part.partName,
                    maxLines: 1, overflow: TextOverflow.ellipsis, style: AppText.body),
                const SizedBox(height: 2),
                Row(
                  children: [
                    Container(
                      width: 10,
                      height: 10,
                      decoration: BoxDecoration(
                        color: swatchColor(part.colorRgb, c.faint),
                        shape: BoxShape.circle,
                        border: Border.all(color: c.line),
                      ),
                    ),
                    const SizedBox(width: AppSpacing.s4),
                    Flexible(
                      child: Text(sub,
                          maxLines: 1, overflow: TextOverflow.ellipsis, style: AppText.caption),
                    ),
                  ],
                ),
              ],
            ),
          ),
          const SizedBox(width: AppSpacing.s8),
          Text(context.l10n.quantityTimes(part.neededQty), style: AppText.label),
        ],
      ),
    );
  }
}

/// Parse a catalog `color_rgb` hex (e.g. `F2CD37`) into a swatch colour.
Color swatchColor(String? rgb, Color fallback) {
  try {
    return Color(int.parse('FF${rgb ?? '808080'}', radix: 16));
  } catch (_) {
    return fallback;
  }
}
